File: other-programming-languages/Node-js/2024/2024-007/2024-007.js
const fs = require('fs');
const path = require('path');

// Helper class to hold test cases
class TestCase {
    constructor(target, numbers) {
        this.target = target;
        this.numbers = numbers;
    }
}

// Evaluate expression left to right with + and *
function evaluateLeftToRight(numbers, ops) {
    let result = numbers[0];
    for (let i = 0; i < ops.length; i++) {
        const op = ops[i];
        const nextNumber = numbers[i + 1];
        switch (op) {
            case '+':
                result += nextNumber;
                break;
            case '*':
                result *= nextNumber;
                break;
            default:
                throw new Error(`Unsupported operator: ${op}`);
        }
    }
    return result;
}

// Evaluate expression left to right with +, *, and ||
function evaluateWithConcat(numbers, ops) {
    let result = numbers[0];
    for (let i = 0; i < ops.length; i++) {
        const op = ops[i];
        const nextNumber = numbers[i + 1];
        switch (op) {
            case '+':
                result += nextNumber;
                break;
            case '*':
                result *= nextNumber;
                break;
            case '||':
                result = parseInt(`${result}${nextNumber}`, 10);
                break;
            default:
                throw new Error(`Unsupported operator: ${op}`);
        }
    }
    return result;
}

// Parse input file into a list of test cases
function parseInput(filePath) {
    const testCases = [];
    const fileContent = fs.readFileSync(filePath, 'utf8');
    const lines = fileContent.split('\n');
    
    lines.forEach(line => {
        if (!line.includes(':')) return;

        const [targetPart, numbersPart] = line.split(':').map(part => part.trim());
        if (!targetPart || !numbersPart) return;

        const target = parseInt(targetPart, 10);
        const numbers = numbersPart.split(/\s+/).map(num => parseInt(num, 10));

        testCases.push(new TestCase(target, numbers));
    });

    return testCases;
}

// Generate all possible operator combinations
function generateOperatorCombinations(operators, length) {
    if (length === 0) return [[]];
    const subCombinations = generateOperatorCombinations(operators, length - 1);
    const combinations = [];
    subCombinations.forEach(sub => {
        operators.forEach(op => {
            combinations.push([...sub, op]);
        });
    });
    return combinations;
}

// Solve Part One using only + and * operators
function solvePartOne(testCases) {
    const operators = ['+', '*'];
    let validTestValuesSum = 0;

    testCases.forEach(testCase => {
        const opsLength = testCase.numbers.length - 1;
        const allOps = generateOperatorCombinations(operators, opsLength);

        const possible = allOps.some(ops => evaluateLeftToRight(testCase.numbers, ops) === testCase.target);
        if (possible) {
            validTestValuesSum += testCase.target;
        }
    });

    return validTestValuesSum;
}

// Solve Part Two using +, *, and || operators
function solvePartTwo(testCases, partOneTime) {
    const operators = ['+', '*', '||'];
    let validTestValuesSum = 0;

    console.log("\nProgress    Interval(s)      Cumulative1(s)    Cumulative2(s)");
    console.log("-------------------------------------------------------------");

    const totalCases = testCases.length;
    const progressInterval = Math.max(Math.floor(totalCases / 10), 1);

    let cumulative1 = 0.0; // Total elapsed time for Part 2 progress
    let cumulative2 = partOneTime; // Cumulative time including Part 1

    const startTime = process.hrtime.bigint();

    testCases.forEach((testCase, index) => {
        const opsLength = testCase.numbers.length - 1;
        const allOps = generateOperatorCombinations(operators, opsLength);

        const possible = allOps.some(ops => evaluateWithConcat(testCase.numbers, ops) === testCase.target);
        if (possible) {
            validTestValuesSum += testCase.target;
        }

        if ((index + 1) % progressInterval === 0 || index + 1 === totalCases) {
            const intervalElapsed = Number(process.hrtime.bigint() - startTime) / 1e9;
            cumulative1 += intervalElapsed;
            cumulative2 = partOneTime + cumulative1;

            console.log(
                `${index + 1}/${totalCases}      ${intervalElapsed.toFixed(9)}   ${cumulative1.toFixed(9)}   ${cumulative2.toFixed(9)}`
            );
        }
    });

    return { validTestValuesSum, cumulative2 };
}

// Main method
function main() {
    const filePath = path.join('\\\\vmware-host\\Shared Folders\\C\\advent-of-code-002\\input-files\\2024\\2024-007\\input.txt');

    // Parse the input
    const testCases = parseInput(filePath);

    // Solve Part One
    console.log("Starting Part 1...");
    const partOneStartTime = process.hrtime.bigint();
    const partOneResult = solvePartOne(testCases);
    const partOneTime = Number(process.hrtime.bigint() - partOneStartTime) / 1e9;
    console.log(`Part 1 finished in ${partOneTime.toFixed(9)} seconds`);
    console.log(`Part 1 Total Calibration Result: ${partOneResult}`);

    // Solve Part Two
    console.log("Starting Part 2...");
    const { validTestValuesSum: partTwoResult, cumulative2: cumulativeTime } = solvePartTwo(testCases, partOneTime);
    const partTwoTime = cumulativeTime - partOneTime;
    console.log(`\nPart 2 finished in ${partTwoTime.toFixed(9)} seconds, cumulative time ${cumulativeTime.toFixed(9)} seconds`);
    console.log(`Part 2 Total Calibration Result: ${partTwoResult}`);
}

// Run the program
main();

File: other-programming-languages/C++-023/2024/2024-007/2024-007.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <chrono>
#include <iomanip>

using namespace std;

struct TestCase {
    long long target;
    vector<long long> numbers;

    TestCase(long long target, vector<long long> numbers) : target(target), numbers(numbers) {}
};

long long evaluateLeftToRight(const vector<long long>& numbers, const vector<string>& ops) {
    long long result = numbers[0];
    for (size_t i = 0; i < ops.size(); ++i) {
        const string& op = ops[i];
        long long nextNumber = numbers[i + 1];
        if (op == "+") {
            result += nextNumber;
        } else if (op == "*") {
            result *= nextNumber;
        } else {
            throw invalid_argument("Unsupported operator: " + op);
        }
    }
    return result;
}

long long evaluateWithConcat(const vector<long long>& numbers, const vector<string>& ops) {
    long long result = numbers[0];
    for (size_t i = 0; i < ops.size(); ++i) {
        const string& op = ops[i];
        long long nextNumber = numbers[i + 1];
        if (op == "+") {
            result += nextNumber;
        } else if (op == "*") {
            result *= nextNumber;
        } else if (op == "||") {
            result = stoll(to_string(result) + to_string(nextNumber));
        } else {
            throw invalid_argument("Unsupported operator: " + op);
        }
    }
    return result;
}

vector<TestCase> parseInput(const string& filePath) {
    vector<TestCase> testCases;
    ifstream file(filePath);
    if (!file.is_open()) {
        cerr << "Error opening file: " << filePath << endl;
        return testCases;
    }

    string line;
    while (getline(file, line)) {
        if (line.find(":") == string::npos) continue;

        size_t colonPos = line.find(":");
        string targetStr = line.substr(0, colonPos);
        string numbersStr = line.substr(colonPos + 1);

        long long target = stoll(targetStr);
        vector<long long> numbers;
        stringstream ss(numbersStr);
        long long num;
        while (ss >> num) {
            numbers.push_back(num);
        }

        testCases.emplace_back(target, numbers);
    }
    return testCases;
}

vector<vector<string>> generateOperatorCombinations(const vector<string>& operators, int length) {
    if (length == 0) return {{}};
    auto subCombinations = generateOperatorCombinations(operators, length - 1);
    vector<vector<string>> combinations;

    for (const auto& sub : subCombinations) {
        for (const string& op : operators) {
            auto newCombination = sub;
            newCombination.push_back(op);
            combinations.push_back(newCombination);
        }
    }
    return combinations;
}

long long solvePartOne(const vector<TestCase>& testCases) {
    vector<string> operators = {"+", "*"};
    long long validTestValuesSum = 0;

    for (const auto& testCase : testCases) {
        bool possible = false;
        int opsLength = testCase.numbers.size() - 1;
        auto allOps = generateOperatorCombinations(operators, opsLength);

        for (const auto& ops : allOps) {
            if (evaluateLeftToRight(testCase.numbers, ops) == testCase.target) {
                possible = true;
                break;
            }
        }
        if (possible) {
            validTestValuesSum += testCase.target;
        }
    }
    return validTestValuesSum;
}

pair<long long, double> solvePartTwo(const vector<TestCase>& testCases, double partOneTime) {
    vector<string> operators = {"+", "*", "||"};
    long long validTestValuesSum = 0;

    int totalCases = testCases.size();
    int progressInterval = max(1, totalCases / 10);

    double cumulative1 = 0.0;
    double cumulative2 = partOneTime;

    cout << "\nProgress    Interval(s)      Cumulative1(s)    Cumulative2(s)\n";
    cout << "-------------------------------------------------------------\n";

    auto start = chrono::high_resolution_clock::now();

    for (int index = 0; index < totalCases; ++index) {
        const auto& testCase = testCases[index];
        bool possible = false;
        int opsLength = testCase.numbers.size() - 1;
        auto allOps = generateOperatorCombinations(operators, opsLength);

        for (const auto& ops : allOps) {
            if (evaluateWithConcat(testCase.numbers, ops) == testCase.target) {
                possible = true;
                break;
            }
        }
        if (possible) {
            validTestValuesSum += testCase.target;
        }

        if ((index + 1) % progressInterval == 0 || (index + 1) == totalCases) {
            auto now = chrono::high_resolution_clock::now();
            double intervalElapsed = chrono::duration<double>(now - start).count();
            cumulative1 += intervalElapsed;
            cumulative2 = partOneTime + cumulative1;

            cout << setw(10) << (index + 1) << "/" << totalCases
                 << setw(20) << fixed << setprecision(9) << intervalElapsed
                 << setw(20) << cumulative1
                 << setw(20) << cumulative2 << endl;

            start = chrono::high_resolution_clock::now();
        }
    }
    return {validTestValuesSum, cumulative2};
}

int main() {
    string filePath = "input.txt"; // Change this to your actual input file location

    auto testCases = parseInput(filePath);

    cout << "Starting Part 1...\n";
    auto partOneStart = chrono::high_resolution_clock::now();
    long long partOneResult = solvePartOne(testCases);
    double partOneTime = chrono::duration<double>(chrono::high_resolution_clock::now() - partOneStart).count();
    cout << "Part 1 finished in " << fixed << setprecision(9) << partOneTime << " seconds\n";
    cout << "Part 1 Total Calibration Result: " << partOneResult << "\n";

    cout << "Starting Part 2...\n";
    auto [partTwoResult, cumulativeTime] = solvePartTwo(testCases, partOneTime);
    double partTwoTime = cumulativeTime - partOneTime;
    cout << "\nPart 2 finished in " << fixed << setprecision(9) << partTwoTime
         << " seconds, cumulative time " << cumulativeTime << " seconds\n";
    cout << "Part 2 Total Calibration Result: " << partTwoResult << "\n";

    return 0;
}
